SIPS/sip-20.md

```diff
@@ -0,0 +1,155 @@
+endowments in order to receive extra information from independent remote sources distributed over the internet.
+
+This proposal provides an improvement that enables one way communication between a Snap and an external data or event
+source.
+The idea is that the external event source sends arbitrary data to the specific controller with listener running
+inside a wallet. The controller then sends the data to the Snap which runs execution of this feature specific RPC
+request handler.
+
+## Specification
+
+> Formal specifications are written in Typescript.
+
+### Language
+
+The key words "MUST", "MUST NOT", "SHOULD", "RECOMMENDED" and "MAY" written in uppercase in this document are to be
+interpreted as described in [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt)
+
+### Snap Manifest
+
+This SIP specifies a permission named `endowment:external-data`.
+
+The new field can be specified as follows in a `snap.manifest.json` file:
+
+```json
+{
+  "initialPermissions": {
+    "endowment:external-data": [
+      {
+        "type": "websocket",
+        "url": "wss://example.com/binary/endpoint",
+        "dataType": "binary"
+      },
+      {
+        "type": "websocket",
+        "url": "wss://example.com/text/endpoint",
+        "dataType": "text"
+      },
+      {
+        "type": "blockchain",
+        "chainId": "eip155:1",
+        "events": [
+          "block"
+        ]
+      }
+    ]
+  }
+}
+```
+
+### Permission caveats
+
+Each caveat object MUST have `type` property that SHOULD be `websocket` or `blockchain`.
+If the `type` is `websocket`then the caveat MUST have `url` and `dataType` properties.
+The `url` MUST be string.
+The `dataType` MUST be string that MUST be `"binary"` or `"text"`.
+The `chainId` MUST be string that represents supported blockchain.
+The `events` SHOULD be an array of strings which represent name of the events for subscription.
+
+- `type` property represents the type of data source.
+- `url` is the WebSocket URL.
+- `dataType` is type of data that WebSocket returns.
+- `chainId` is the blockchain ID.
+- `events` is the array of events to subscribe.
+
+The caveats MUST NOT have duplicate objects with the same `url` or `chainId` properties.
+
+The `url` MUST start with `wss://` which is a protocol indicator for secure WebSocket connection.
+
+### Snap implementation (request handler)
+
+This SIP proposes the following RPC request handler to execute implemented functionality for external data events:
+
+#### OnExternalDataHandler
+
+Example:
+```typescript
+import { OnExternalDataHandler } from '@metamask/snaps-types';
+import { assert } from '@metamask/utils';
+
+export const onExternalData: OnExternalDataHandler = ({ id, type, data, source }) => {
+  assert(type === 'websocket'); // `data` is inferred as `WebSocketData`.
+  assert(data.type === 'text'); // `message` is inferred as `string`.
+
+  // Now the Snap can do whatever is needed with the message.
+  const json = JSON.parse(data.message);
+};
+```
```

Comment:
I would move this to an `#### Example` section again.

Reply:
Updated.